compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -151,19 +185,33 @@ impl<'r, 'ast> Visitor<'ast> for AccessLevelsVisitor<'ast, 'r> {
                 self.set_bindings_access_level(def_id);
                 for variant in variants {
                     let variant_def_id = self.r.local_def_id(variant.id);
-                    let variant_level = self.r.access_levels.get_access_level(variant_def_id);
                     for field in variant.data.fields() {
-                        self.set_access_level(field.id, variant_level);
+                        let field_def_id = self.r.local_def_id(field.id);
+                        let vis = self.r.visibilities.get(&field_def_id).unwrap().clone();
+                        let field_span = field.span;
+                        self.update_effective_vis(
+                            field_def_id,
+                            vis,
+                            variant_def_id,
+                            AccessLevel::Public,
+                            field_span,
+                        );
                     }
                 }
             }
 
             ast::ItemKind::Struct(ref def, _) | ast::ItemKind::Union(ref def, _) => {
-                let inherited_level = self.r.access_levels.get_access_level(def_id);
                 for field in def.fields() {
-                    if field.vis.kind.is_pub() {
-                        self.set_access_level(field.id, inherited_level);
-                    }
+                    let field_def_id = self.r.local_def_id(field.id);
+                    let vis = self.r.visibilities.get(&field_def_id).unwrap();
+                    let field_span = field.span;
+                    self.update_effective_vis(
+                        field_def_id,
+                        *vis,
+                        def_id,
+                        AccessLevel::Public,
+                        field_span,
```

Comment:
```suggestion
                        field.span,
```
Nit: some unnecessary span variables here and above.